Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs

```diff
@@ -111,6 +116,18 @@ ILoggerFactory loggerFactory
         notificationPublisher.SendStatusNotification(textDocument, CompilationStatus.ParsingFailed);
         return CreateDocumentWithEmptySymbolTable(loggerFactory.CreateLogger<SymbolTable>(), textDocument, errorReporter, program, loadCanceled: false);
       }
+
+      // We load plugins at most once.
```

Comment:
I must have missed this before. Plugin loading should happen somewhere during the initialization of the server; thus, I strongly believe it's the wrong place to do that inside the document loading procedure.
By moving it somewhere to the initialization you no longer have to check if the plugins have been loaded before. While it's not relevant yet, you also won't have to deal with concurrency.

Reply:
Thanks a lot for the tip. I was able to also remove the check that plugins were not loaded yet.